apiary-metastore-metrics/src/main/java/com/expedia/apiary/extensions/metastore/metrics/CodahaleMetrics.java

```diff
@@ -0,0 +1,503 @@
+    Gauge gauge = new Gauge() {
+      @Override
+      public Object getValue() {
+        return variable.getValue();
+      }
+    };
+    addGaugeInternal(name, gauge);
+  }
+
+  @Override
+  public void addRatio(String name, MetricsVariable<Integer> numerator,
+                           MetricsVariable<Integer> denominator) {
+    Preconditions.checkArgument(numerator != null, "Numerator must not be null");
+    Preconditions.checkArgument(denominator != null, "Denominator must not be null");
+
+    MetricVariableRatioGauge gauge = new MetricVariableRatioGauge(numerator, denominator);
+    addGaugeInternal(name, gauge);
+  }
+
+  private void addGaugeInternal(String name, Gauge gauge) {
+    try {
+      gaugesLock.lock();
+      gauges.put(name, gauge);
+      // Metrics throws an Exception if we don't do this when the key already exists
+      if (metricRegistry.getGauges().containsKey(name)) {
+        LOGGER.warn("A Gauge with name [" + name + "] already exists. "
+            + " The old gauge will be overwritten, but this is not recommended");
+        metricRegistry.remove(name);
+      }
+      metricRegistry.register(name, gauge);
+    } finally {
+      gaugesLock.unlock();
+    }
+  }
+
+  @Override
+  public void markMeter(String name) {
+    String key = name;
+    try {
+      metersLock.lock();
+      Meter meter = meters.get(name);
+      meter.mark();
+    } catch (ExecutionException e) {
+      throw new IllegalStateException("Error retrieving meter " + name
+          + " from the metric registry ", e);
+    } finally {
+      metersLock.unlock();
+    }
+  }
+
+  // This method is necessary to synchronize lazy-creation to the timers.
+  private Timer getTimer(String name) {
+    String key = name;
+    try {
+      timersLock.lock();
+      Timer timer = timers.get(key);
+      return timer;
+    } catch (ExecutionException e) {
+      throw new IllegalStateException("Error retrieving timer " + name
+          + " from the metric registry ", e);
+    } finally {
+      timersLock.unlock();
+    }
+  }
+
+  private void registerAll(String prefix, MetricSet metricSet) {
+    for (Map.Entry<String, Metric> entry : metricSet.getMetrics().entrySet()) {
+      if (entry.getValue() instanceof MetricSet) {
+        registerAll(prefix + "." + entry.getKey(), (MetricSet) entry.getValue());
+      } else {
+        metricRegistry.register(prefix + "." + entry.getKey(), entry.getValue());
+      }
+    }
+  }
+
+  @VisibleForTesting
+  public MetricRegistry getMetricRegistry() {
+    return metricRegistry;
+  }
+
+  @VisibleForTesting
+  public String dumpJson() throws Exception {
+    ObjectMapper jsonMapper = new ObjectMapper().registerModule(
+      new MetricsModule(TimeUnit.MILLISECONDS, TimeUnit.MILLISECONDS, false));
+    return jsonMapper.writerWithDefaultPrettyPrinter().writeValueAsString(metricRegistry);
+  }
+
+  private void initCloudwatchReporting() {
```

Comment:
Put a comment here to indicate this is a change we've added.

Reply:
Also, wouldn't it be better to move this method into the switch statement below so it's only activated if "reportingSet" is CLOUD_WATCH? Any idea how we can control the values?

Reply:
This will require us to pull more code from hive project.
need to implement this https://github.com/apache/hive/blob/branch-2.3/common/src/java/org/apache/hadoop/hive/common/metrics/metrics2/MetricsReporting.java and add CLOUD_WATCH, means further changes to CodahaleMetrics.java.

I think this is only important if we anticipate to use CLOUD_WATCH and other reporters simultaneously.